doc/oopsctl.8.in
.TH OOPSCTL 8 "Sep 2000" "Oopsctl"
.SH NAME
oopsctl \- oops HTTP caching proxy server controller
.SH SYNOPSIS
.B @OOPS_SBINDIR@/oopsctl
.RB [-s\ pathtosocket]\ [command]
.SH DESCRIPTION
.B oopsctl
is intended to controll the \fBoops\fP proxy server
.SH OPTIONS
.IP \fB-s\ file\fP
Where is the unix socket file to send commands to the \fBoops\fP server?
Generally that is
.RI @OOPS_LOCALSTATEDIR@/oopsctl
.IP \fBhelp\fP
get help
.IP \fBstat\fP
get stat
.IP \fBhtmlstat\fP
get stat in html format
.IP \fBchkconfig\fP
check config file
.IP \fBreconfigure\fP
re-read config file
.IP \fBshutdown\fP
shutdown oops
.IP \fBstop\fP
same as the \fBshutdown\fP
.IP \fBrotate\fP
rotate logs
.IP \fBverbosity=LVL\fP
set verbosity (like \fB-x\ LVL\fP)
.IP \fBstart\fP
start oops (same as \fB@OOPS_SBINDIR@/oops\ -c\ @OOPS_SYSCONFDIR@/oops.cfg\fP)
.PP
.SH AUTHOR
Mainly Igor Khasilev <igor@paco.net>


ChangeLog
version 1.5.23  - netflow module
                - status 404 cached with configurable expire
                  more close to rfc2616 in IMS processing
version 1.5.22
		- wccp v2. Real raw devices support (aligned IO).
		  new swapout algorithm. Configurable option for speed
		  between server and proxy (with client speed(default) or
		  with full speed). More memory control. Performance
		  emprovements, bug fixes, e.t.c.
version 1.5.18
		 - method CONNECT ignored connect_from.
	24.06.01 - Initial OpenBSD support. (ai)

version 1.5.17
	26.04.01 - connect_from for groups. (ai)
	12.04.01 - acces_log now htmlize path.
		   accel.c: host from absolute uri have precedence over
			Host:
		   ssl works with parent
        05.04.01 - configure.in fix for SUN Forte Workshop 6. (ai)
	05.03.01 - Cookies, Parent Authentication.
	03.03.01 - 'allow' rules for redir.c, fastredir.c
	13.02.01 - Fixed memory-corruption bug in ftp_utils.c
		   Fixed error (Host: was not inserted if need in requests
		   when force_http11 option commented out).
		   Fixed bug: Proxy-Authorization now is hop-by-hop
version 1.5.9
	30.01.01 - redir and fastredir can use different configs
		   like redir/1 redir/2 ...
		   local_networls was not handled properly if ICP is used - one
		   more fix.
	28.01.01 - icp processing use hash now.
		   rand_r used in request numbers in icp requests.
		   hash.c rewriten (dst_domain_acl used this interface)
version 1.5.8
	23.01.01 - icp requests now processed in workers.
	20.01.01 - fixed bug in call to pread.
		   worker threads returns after timeout.
	12.01.01 - fixed bug in acl.c (several networks_acl for group didn't work
		   moved to new workq - workers should live with timeouts -
		   will not run lot of processes under linux.
version 1.5.6
	29.12.00 - username acl
	28.12.00 - \n removed ftom some acl's when read from file
		   local_networls was not handled properly if ICP is used
		   no log messages 'icp so = -1' if icp_port  0
version 1.5.4
	18.12.00 - more sprintf/strncpy fixes
		   memory leaks in accel.c, bug in custom_log.c
		   bug in parse_named_acl_data for SRC_IP acl.
version 1.5.0
	11.12.00 - fixed memory/security bug if ftp_utils (allocated less
		   memory in heap thaen required for htmlized_something)
	23.11.00 - check for directio()
	17.11.00 - redir.c can rewrite url or send Location:
		   (otion mode in redir config)
		   fixed very OLD bug in buid_destination:redir.c, which
		   caused SIGSEGV when you use long destinations in redir rules
	15.11.00 - fail-safe support for non-icp peers.
	14.11.00 - always_check_freshness on object acl basis
		   acl type "content_type"
	11.11.00 - migrate to pthread_rwlock_*. (ai)
		   fixed build problem for AIX. (ai)
	08.11.00 - ungzip gzipped content when send to non-gzip capable useragent
		   original_host and original_path in redir's
	30.10.00 - performance bug in storage.c
		   updated passwd_pgsql, passwd_mysql modules
		   IMPORTANT: Changed magic in storages!
	22.10.00 - passwd_pgsql, passwd_mysql modules
		   by "Ivan B. Yelnikov" <bahek@khspu.ru>
		   redir.c can send internal gifs (currently only nospam1x1)
		   instead of redirects.
	20.10.00 - fix old bug in transparent.c
		   by "Alexander Savelyev" <fano@fano.com.ua>
	16.10.00 - per_ip_bw and per_ip_conn limits - limit
		   bandwidth and connections nubmer for any host.
	13.10.00 - fixed bug in bind_acl (vova@mics.msu.su)
		   per_ip_bw - per ip address bandwidth contol
	08.10.00 - fixed bug int icp (if peer name can't be resolved, send icp
		   rq to 0.0.0.0)
		   fixed bug: traffic load was not counted in continue_load
		   fixed in oopsctl_mod htmlstat descriptions for DB modules.
version 1.4.22
	02.10.00 - manpages (contributed by Tamas SZERB <toma@rulez.org>)
	27.09.00 - Fixed bug in clean_disc.c (core during expire with GigaBASE)
		   Fixed bug with oops.pid reopen (oopsctl reconfigure effectively
		   do oopsctl stop). It was introduced during FreeBSD port prepare.
	21.09.00 - oopsctl verbosity=LVL
	14.09.00 - fixed security fix ([a:b] in badports procssed incorrectly)
	13.09.00 - configure can produce "FreeBSD port"-ready Makefile
		   fixed bug in lang.c (by Peter S. Voronov)
	16.08.00 - parse_myports(): bind problem for FreeBSD solved.
	14.08.00 - parse_myports(): bind problem report.
	11.08.00 - fixed bug in refresh_pattern percent calculation
		   Mark expired documents as dead during leave.
	10.08.00 - GigaBASE seems to to work now.
	02.08.00 - compile without BerkeleyDB
		   peer_access using ACL option
	01.08.00 - fixed bug when icp-peer have no allow and deny statements
version 1.4.6
	09.06.00 - Authorisation with parent/sibling
		   --with-malloc option in configure (for exapmle
		   I use --with-malloc=-lptmalloc under Solaris. ptmalloc
		   have to be compiled and installed separately. Take
		   it, please, from http://malloc.de/).
	08.06.00 - 'Range:' header and restart(resume) for ftp and http
version 1.4.5
	24.05.00 - %U in custom_log
	21.05.00 - buffering system for log and accesslog rewritten.
		   err.c - for ERR_BAD_URL show succesfully parsed fields.
		   pthread_attr_setstacksize(64k) under solaris.
		   some minor bugs fixed (^C for FreeBSD, HTTP 0.9 docs,...)(ik)
		   configure --enable-db-locks for DB 2.x - turn off top-level
		   locks for some DB operations. Can emprove speed (tested, but
		   still experimental)
	03.05.00 - configure --with-regexp[=(builtin|pcre|system)] [system] (ai)
	29.04.00 - Initial WIN32 support. (ai)
	27.04.00 - module fastredir by john gladkih
		   redir based on strcasestr, which is much faster then
		   regex, but not so flexible.
version.1.4
	24.04.00 - some daemon() related stuff fixed(ik)
	18.04.00 - OPTIONS method added. (ik)
		   FreeBSD 4.0 fix for signals(siginterrupt). (ik)
        12.04.00 - The `size auto' parameter is now working on raw devices in
	           BSDI, FreeBSD and Linux. (ai)
        11.04.00 - Initial BSDI support. (ai)

src/modules/oopsctl.c
/*
Copyright (C) 1999 Igor Khasilev, igor@paco.net

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

*/

#define		NO_NEED_XMALLOC		1
#define		_LIB_H_INCLUDED_	1

#include	"../oops.h"

static	char	path[MAXPATHLEN];
static	char	cpath[MAXPATHLEN];
static	char	*command[20];

int
main(int argc, char **argv)
{
struct	sockaddr_un	sun_addr;
int			oopsctl_so;
char			answer[1024];
int			alen,i,first_arg = 1;
char 			cmdarg[1024];

    for (i=0; i< 20; i++) { *(i+command) = NULL;}
    *command = strdup("help");
    *(2 + command) = strdup("-c");
    snprintf(cmdarg, sizeof(cmdarg)-1, "%s", OOPS_CFG);
    *(3 + command) = strdup(cmdarg);
    i = 4;
    path[0] = 0;
    argc--;argv++;
    while ( argc && *argv) {
	if ( !strncasecmp(*argv, "-s", 2) ) {
	    /* path to socket */
	    if ( strlen(*argv) <= 2 ) {
		/* mustbe next argv */
		argv++;argc--;
		strncpy(path, *argv, sizeof(path)-1);
	    } else
		strncpy(path, (*argv+2), sizeof(path));
	    path[sizeof(path)-1] = 0;
        } else if (!strncasecmp(*argv, "-c", 2)) {
            /* path to config file */
            if ( strlen(*argv) <= 2 ) {
                /* mustbe next argv */
                argv++;argc--;
		strncpy(cpath, *argv, sizeof(cpath)-1);
	    } else
		strncpy(cpath, (*argv+2), sizeof(cpath)-1);
	    cpath[sizeof(cpath)-1] = 0;
            *(3 + command) = cpath;
	} else {
            if ( i < 19 ) {
 	      if ( first_arg ) {
		*command = *argv;
		first_arg = 0 ;
	      } else {
	        *(i+command) = *argv;
                i++;
	      }
            } else {
              printf("Too many argc !\n\n");
            }
	}
	argc--; argv++;
    }
    if ( !path[0] ) {
	strncpy(path, OOPS_LOCALSTATEDIR, sizeof(path)-1);
	strncat(path, "/oopsctl", sizeof(path)-strlen(path)-1);
	path[sizeof(path)-1] = 0;
    }
    if ( !strcasecmp(*command, "help") ) {
	printf("oopsctl [-s pathtosocket] [command]\n");
	printf("Commands:\n");
	printf("help		- get help\n");
	printf("stat		- get stat\n");
	printf("htmlstat	- get stat in html format\n");
	printf("chkconfig	- check config file\n");
	printf("reconfigure	- re-read config file\n");
	printf("graceful	- re-read config file without storage checking\n");
	printf("shutdown(stop)	- shutdown oops\n");
	printf("rotate		- rotate logs\n");
	printf("verbosity=LVL	- set verbosity (like -x LVL)\n");
	printf("requests	- print current requests in service\n");
	printf("start		- start oops (same as %s/oops -c %s)\n", OOPS_SBINDIR,OOPS_CFG);
	exit(0);
    } else
    if ( !strcasecmp(*command, "start") ) {
	pid_t	child;
	char	cmdpath[1024];

	snprintf(cmdpath, sizeof(cmdpath)-1, "%s/oops", OOPS_SBINDIR);
	*(1+command) = cmdpath;
        chdir(OOPS_HOME);
	child = fork();
	switch(child) {
	case(-1):
		printf("oopsctl: Can't start child: %s\n", strerror(ERRNO));
		exit(1);
		break;
	case(0):
		i = 4;
		if (*(command + 4)) {
		 printf ("args: ");
		 while ( *(command + i) ) printf ("%s ",*(command + i++));
		 printf ("\n");
		}
		execv(cmdpath, command+1);
		printf("oopsctl: Can't execute: %s\n", strerror(ERRNO));
	default:
		exit(0);
	}
    }
    if ( !strcasecmp(*command, "chkconfig") ) {
	char	cmdpath[1024], cmdarg[1024];
	snprintf(cmdpath, sizeof(cmdpath)-1, "%s/oops", OOPS_SBINDIR);
	snprintf(cmdarg, sizeof(cmdarg)-1, "-C%s", OOPS_CFG);
	*(1+command) = cmdpath;
	*(2+command) = cmdarg;
	execv(cmdpath, command+1);
	printf("oopsctl: Can't execute: %s\n", strerror(ERRNO));
    }
    if ( !strcasecmp(*command, "reconfigure") ) {
	char	cmdpath[1024], cmdarg[1024];
	pid_t	child;
	int	stat;

	/* first check if config is OK */
	snprintf(cmdpath, sizeof(cmdpath)-1, "%s/oops", OOPS_SBINDIR);
	snprintf(cmdarg, sizeof(cmdarg)-1, "-C%s", OOPS_CFG);
	*(1+command) = cmdpath;
	*(2+command) = cmdarg;
	child = fork();
	switch (child) {
	case -1:
		printf("oopsctl: Can't start child: %s\n", strerror(ERRNO));
		exit(1);
		break;
	case 0:
		execv(cmdpath, command+1);
		printf("oopsctl: Can't execute: %s\n", strerror(ERRNO));
	default:
		/* wait for child */
		waitpid((pid_t)-1, &stat, 0);
		if ( WIFEXITED(stat) && !WEXITSTATUS(stat) ) { /* ok */
		    *(1+command) = NULL;
		    goto srv_conn;
		}
		printf("oopsctl: Check config failed: exitcode: %d\n", WEXITSTATUS(stat));
		exit(1);
	}
    }
    if ( !strcasecmp(*command,"stat") || 
	 !strcasecmp(*command,"htmlstat") ||
	 !strcasecmp(*command,"reconfigure") ||
	 !strcasecmp(*command,"graceful") ||
	 !strcasecmp(*command,"flush") ||
	 !strcasecmp(*command,"rotate") ||
	 !strcasecmp(*command,"stop") ||
	 !strncasecmp(*command,"verbosity=", 10) ||
	 !strncasecmp(*command,"requests", 8) ||
	 !strcasecmp(*command,"shutdown")
        ) {

srv_conn:
	/* connecting to server */
	oopsctl_so = socket(AF_UNIX, SOCK_STREAM, 0);
	if ( oopsctl_so == -1 ) {
	    printf("oopsctl: socket(): %s\n", strerror(ERRNO));
	    exit(1);
        }
        bzero(&sun_addr, sizeof(sun_addr));
        sun_addr.sun_family = AF_UNIX;
        strncpy(sun_addr.sun_path, path, sizeof(sun_addr.sun_path)-1);
        if ( connect(oopsctl_so, (struct sockaddr*)&sun_addr, sizeof(sun_addr)) ) {
	    printf("oopsctl: connect(): %s\n", strerror(ERRNO));
	    exit(1);
        }
        write(oopsctl_so, *command, strlen(*command));
        write(oopsctl_so, "\n", 1);
        fflush(stdout);
        while ( (alen = read(oopsctl_so, answer, sizeof(answer))) > 0 ) {
	    write(1, answer, alen);
        }

    } else
	printf("oopsctl: Unknown command %s\n", *command);

    return(0);
}
